esmvalcore/_recipe.py

```diff
@@ -387,37 +386,57 @@ def _add_fxvar_keys(fx_var_dict, variable):
 
 def _get_correct_fx_file(variable, fx_varname, config_user):
     """Get fx files (searching all possible mips)."""
-    # TODO: allow user to specify certain mip if desired
+    # make it a dict
+    if not isinstance(fx_varname, dict):
+        fx_varname = {'short_name': fx_varname}
+
+    # get some parameters
+    user_fx_mip = fx_varname.get('mip')
+    user_fx_experiment = fx_varname.get('exp')
+    fx_varname = fx_varname.get('short_name')
+
+    # assemble info from master variable
     var = dict(variable)
     var_project = variable['project']
     cmor_table = CMOR_TABLES[var_project]
+    valid_fx_vars = []
 
     # Get all fx-related mips ('fx' always first, original mip last)
     fx_mips = ['fx']
     fx_mips.extend(
         [key for key in cmor_table.tables if 'fx' in key and key != 'fx'])
     fx_mips.append(variable['mip'])
 
+    # force only the mip declared by user
+    if user_fx_mip:
+        fx_mips = [
+            user_fx_mip,
+        ]
+
     # Search all mips for available variables
+    # priority goes to user specified mip if available
     searched_mips = []
+    fx_files = []
     for fx_mip in fx_mips:
         fx_variable = cmor_table.get_variable(fx_mip, fx_varname)
         if fx_variable is not None:
             searched_mips.append(fx_mip)
-            fx_var = _add_fxvar_keys(
-                {'short_name': fx_varname, 'mip': fx_mip}, var)
-            logger.debug("For CMIP6 fx variable '%s', found table '%s'",
-                         fx_varname, fx_mip)
+            fx_var = _add_fxvar_keys({
+                'short_name': fx_varname,
+                'mip': fx_mip
+            }, var)
+            if user_fx_experiment:
+                fx_var['exp'] = user_fx_experiment
```

Comment:
I think replacing all this by

```python
fx_var = _add_fxvar_keys(fx_varname, var)
```

and simplifying `_add_fxvar_keys` by

```python
def _add_fxvar_keys(fx_var_dict, variable):
    """Add keys specific to fx variable to use get_input_filelist."""
    fx_variable = dict(variable)
    fx_variable.update(fx_var_dict)
    fx_variable['variable_group'] = fx_var_dict['short_name']
    if fx_variable['project'] == 'CMIP5':
        fx_variable['ensemble'] = 'r0i0p0'
    _add_cmor_info(fx_variable, override=True)
return fx_variable
```

should also the job + allows using other options than `exp` and `mip` in the `dict` for fx_files. You can also remove l.394 to 396 then.

Reply:
no that's not working(tested until now, since you gave me the suggestion) because there's the mips that need to be changed for the fx variables otherwise the cmor table loading fails

Reply:
Ah my bad, sorry. You need to add the `mip` to the `fx_varname` dictionary beforehand, e.g.

```python
fx_varname['mip'] = fx_mip
fx_var = _add_fxvar_keys(fx_varname, var)
```

then it should work.

Reply:
it does work but the last test fails:
```
def test_fx_var_invalid_project(tmp_path, patched_datafinder, config_user):
    content = dedent("""
        preprocessors:
          preproc:
           area_statistics:
             operator: mean
             fx_files: ['areacella']

        diagnostics:
          diagnostic_name:
            variables:
              tas:
                preprocessor: preproc
                project: EMAC
                mip: Amon
                exp: historical
                start_year: 2000
                end_year: 2005
                ensemble: r1i1p1f1
                grid: gn
                additional_datasets:
                  - {dataset: CanESM5}
            scripts: null
        """)
    msg = 'Project EMAC not supported with fx variables'
    with pytest.raises(RecipeError) as rec_err_exp:
        get_recipe(tmp_path, content, config_user)
        assert msg in rec_err_exp
```
I cant see anywhere the expected error - where is it?

Reply:
I removed the test since it was dodgy, pls put it back in (if in a working condition) but I dont see a need for it tho

Reply:
cool dis fixed now